modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/PartitionListener.java

```diff
@@ -227,72 +166,52 @@ private void handleUpdateCommand(UpdateCommand cmd, long commandIndex) {
 
         txsPendingRowIds.computeIfAbsent(txId, entry -> new ConcurrentHashSet<>()).add(rowId);
 
-        storage.lastAppliedIndex(commandIndex);
+        if (row == null) {
+            // Remove entry.
+            txsRemovedKeys.computeIfAbsent(txId, entry -> new ConcurrentHashSet<>()).add(rowId);
```

Comment:
PartitionLisener is single threaded,  so there's no sense in using inner concurrent structures inside it. Please update
- txsInsertedKeys
- txsRemovedKeys
- txsPendingRowIds
to be simple Maps instead of Concurrent ones along with using HashSets instead of ConcurrentHashSet inside them.